### PR TITLE
#18 Deploy with cmake

### PR DESCRIPTION
**Pull request by kreuzberger**
Deploy the libraries, namespacing, includes from cmake.

Makes it possible to include the "installed" directory with

```
find_package(
    CwTCucumber
    2.0.0
    REQUIRED
    PATHS "${CONAN_CWT-CUCUMBER_ROOT}"
    NO_DEFAULT_PATH
)
```

In conanfile.py the package part is reduced to a simple `cmake --build . --target install`.
A fine tuned install in conanfile.py could be
```
        # CMake(self).install()
        command_line = [f"-DCMAKE_INSTALL_PREFIX={self.package_folder}",
                        f"-DCMAKE_INSTALL_COMPONENT=cucumber"]

        self.run(
            "cmake %s -P %s" % (" ".join(command_line), self.build_folder + "/cmake_install.cmake")
        )
```
This would not install box executable

To let the conanfile build and install propery without copying from build to package directory, configure PROPER CMAKE_INSTALL_PREFIX (rest is personal preference, i use ninja for linux/windows, could be make/vs/jom whatever)
```
   def build(self):
        cmake = CMake(self)
        if self.should_configure:
            command_line = [
                f"-DCMAKE_INSTALL_PREFIX={self.package_folder}",
                f"-DCMAKE_BUILD_TYPE={self.settings.build_type}"
            ]
            self.run(
                'cmake -G Ninja "-S %s" "-B %s" %s'
                % (self.source_folder, self.build_folder, " ".join(command_line))
            )
        if self.should_build:
            self.run("cmake --build . --target all")
```

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -8,10 +8,11 @@ if (NOT CMAKE_BUILD_TYPE)
...
 endif()
 
-set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
-set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
-set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+include(GNUInstallDirs)
 
+# set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
+# set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
+# set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
```
Comment (ThoSe1990) [inline CMakeLists.txt:15]: I use this directories a lot during dev. We can toggle between GNUInstallDirs with an option e.g. USE_GNU_INSTALL_DIRS or PACKAGE_BUILD

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -16,22 +16,73 @@ set(cucumber-src
...
+
+
+# install(
+#   TARGETS cucumber
+#   COMPONENT cucumber
+# )
+# install(
+#   TARGETS cucumber-no-main
+#   COMPONENT cucumber
+# )
```
Comment (ThoSe1990) [inline src/CMakeLists.txt:38]: delete, no comments please

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -16,22 +16,73 @@ set(cucumber-src
...
+set(LIBRARY_SYSTEM_NAME "${CMAKE_HOST_SYSTEM_NAME}")
+
+configure_package_config_file(
+    CwTCucumberConfig.cmake.in ${CMAKE_CURRENT_BINARY_DIR}/CwTCucumberConfig.cmake
+    INSTALL_DESTINATION lib/cwtcucumber/cmake
+    PATH_VARS INCLUDE_INSTALL_DIR LIBRARY_INSTALL_DIR LIBRARY_VERSION LIBRARY_SYSTEM_NAME
+)
+
+write_basic_package_version_file(
+    ${CMAKE_CURRENT_BINARY_DIR}/CwTCucumberConfigVersion.cmake
```
Comment (ThoSe1990) [inline src/CMakeLists.txt:54]: I never used the naving convention CwT, please use here cwt-... so it is consistent with the conan package from the package repo

**Comment by ThoSe1990** [comment]
So first of all thanks for the effort here. As mentioned before I never used CMake packages in that way. but we can introduce it since for the existing project the installed command is not really used. 

It's not really clear to me which `conanfile.py` you mean (I believe it is one from you). But we can merge it after you went through my comments. Most important to me is the naming convention and that we can toggle the output directories

**Comment by kreuzberger** [comment]
Ok, read the comments and would commit after my further investigations. With the current branch i am able to use it via conan in my code and evalutate handling and useage. so don't wonder why the changes are not done immediatly.

**Comment by ThoSe1990** [comment]
yh sure, but i'm wondering why don't you use the recipe which I wrote to create and consume this package? 

https://github.com/ThoSe1990/cwt-cucumber-conan

**Comment by kreuzberger** [comment]
2 major reasons:

* still not on conan 2.0
* building / working without internet access, therefore different source export/import handling.

File: CMakeLists.txt (hunk 1 above)
Comment (kreuzberger) [reply CMakeLists.txt]: The usage of GnuInstallDirs is mandatory for the Package export, but does not affect the output directory settings used. So both could be used without any conflicts. Moved GNUInstallDirs to the package definition (src/CMakeLists.txt) and uncommented the old code

File: src/CMakeLists.txt (hunk 2 above)
Comment (kreuzberger) [reply src/CMakeLists.txt]: deleted

**Comment by ThoSe1990** [comment]
please rename it to `cwt-cucumber` in the cmakelists and the configuration file, then we're good to go :)

**Comment by ThoSe1990** [comment]
> please rename it to `cwt-cucumber` in the cmakelists and the configuration file, then we're good to go :)

I mean the config file: `src/cwtcucumber-config.cmake.in` to `src/cwt-cucumber-config.cmake.in` and of course their occurences in the cmakelists

**Pull request merged**